File: Java/src/main/java/QuestionBank/Solution701_800/Solution766.java
package QuestionBank.Solution701_800;

/**
 * @author windows
 */
public class Solution766 {
    public boolean isToeplitzMatrix(int[][] matrix) {
        int i = 0;
        int j = matrix[0].length - 2;
        for (; j >= 0; j--) {
            int tempI = i + 1;
            int tempJ = j + 1;
            while (tempJ < matrix[0].length && tempI < matrix.length){
                if (matrix[tempI][tempJ] != matrix[i][j]){
                    return false;
                }
                tempI++;
                tempJ++;
            }
        }
        i = 1;
        j = 0;
        for (; i < matrix.length; i++) {
            int tempI = i + 1;
            int tempJ = j + 1;
            while (tempJ < matrix[0].length && tempI < matrix.length){
                if (matrix[tempI][tempJ] != matrix[i][j]){
                    return false;
                }
                tempI++;
                tempJ++;
            }
        }
        return true;
    }


    public static void main(String[] args) {
        Solution766 solution766 = new Solution766();
        System.out.println(solution766.isToeplitzMatrix(new int[][]{{1, 2, 3, 4}, {5, 1, 2, 3}, {9, 5, 1, 2}}));
    }
}

File: C++/src/Solution701_800/Solution766.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution766.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

bool Solution::isToeplitzMatrix(vector<vector<int>> &matrix) {
    int m = matrix.size(), n = matrix[0].size();
    int j = 0;
    for (int i = 0; i < n; ++i) {
        int x = j + 1, y = i + 1;
        while(x < m && y < n){
            if (matrix[x][y] == matrix[j][i]){
                x++;
                y++;
            } else{
                return false;
            }
        }
    }
    j = 0;
    for (int i = 1; i < m; ++i) {
        int x = i + 1, y = j + 1;
        while(x < m && y < n){
            if (matrix[x][y] == matrix[i][j]){
                x++;
                y++;
            } else{
                return false;
            }
        }
    }
    return true;
}
